core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java

```diff
@@ -0,0 +1,192 @@
+
+        String fromImage = buildConfiguration.getFrom();
+        String targetImage = buildConfiguration.getTargetImage();
+        Credential credential = buildConfiguration.getCredential();
+        Map<String, String> envMap  = buildConfiguration.getEnvMap();
+        List<String> portList = buildConfiguration.getPorts();
+        Set<Port> portSet = getPortSet(portList);
+        String  outputDir = buildConfiguration.getOutputDir();
+        String targetDir = buildConfiguration.getTargetDir();
+        Path fatJar = buildConfiguration.getFatJar();
+
+        List<String> entrypointList = new ArrayList<>();
+        if(buildConfiguration.getEntryPoint() != null) {
+            entrypointList = buildConfiguration.getEntryPoint().asStrings();
+        }
+
+        buildImage(fromImage, targetImage, envMap, credential, portSet, fatJar, entrypointList, targetDir, outputDir, log);
+    }
+
+    private static Set<Port> getPortSet(List<String> ports) {
+
+        Set<Port> portSet = new HashSet<Port>();
+        for(String port : ports) {
+            portSet.add(Port.tcp(Integer.parseInt(port)));
+        }
+
+        return portSet;
+    }
+
+    protected static JibContainer buildImage(String baseImage, String targetImage, Map<String, String> envMap, Credential credential, Set<Port> portSet, Path fatJar, List<String> entrypointList, String targetDir, String outputDir, Logger log) throws InvalidImageReferenceException {
+
+        String username = "";
+        String password = "";
+        String imageTarName = ImageReference.parse(targetImage).getRepository().concat(".tar");
+
+        JibContainerBuilder contBuild = Jib.from(baseImage);
+
+        if (envMap != null) {
+            contBuild = contBuild.setEnvironment(envMap);
+        }
+
+        if (portSet != null) {
+            contBuild = contBuild.setExposedPorts(portSet);
+        }
+
+
+        if (fatJar != null) {
+            String fatJarName = fatJar.getFileName().toString();
+            String jarPath = targetDir + "/" + (fatJarName.isEmpty() ? DEFAULT_JAR_NAME: fatJarName);
+            contBuild = contBuild
+                    .addLayer(LayerConfiguration.builder().addEntry(fatJar, AbsoluteUnixPath.get(jarPath)).build());
+        }
+
+        if(!entrypointList.isEmpty()) {
+            contBuild = contBuild.setEntrypoint(entrypointList);
+        }
+
+        if (credential != null) {
+            username = credential.getUsername();
+            password = credential.getPassword();
+        }
+
+        RegistryImage registryImage = RegistryImage.named(targetImage);
+        TarImage tarImage = TarImage.named(targetImage).saveTo(Paths.get(outputDir + "/" + imageTarName));
+
+        try {
+            Containerizer containerizer = Containerizer.to(registryImage.addCredential(username, password));
+            JibContainer jibRegistryImageContainer = contBuild.containerize(containerizer);
+            log.info("Image %s successfully built and pushed.", targetImage);
```

Comment:
In my opinion, the default behavior should be generation of image locally. Why are we doing that as a fallback? Shouldn't image push step go in `PushMojo` ?

Reply:
Jib auto pushes the images to registry if the creds are verified!

Reply:
That's what a RegistryImage is meant to do in JIB, that's why they have this type of image!

Reply:
I don't like the idea of pushing image during build goal as it does not align with mojo's standard functionality. I would prefer finding some offline solution if possible.

Reply:
We have the offline solution, Jib supports three kinds
1.DockerDaemonImage - We don't need it as we are making it daemonless.
2.RegistryImage - It will do auto push that's Jib's functionality.
3.TarImage - This option is already added.

Reply:
Does the user get to choose these build options? If not, we can provide some way to override the default behavior. The default behavior should be building tarball.